### PR TITLE
[RSDK-6525] Start pulling the data-reading code out of the data-interpretation code for NMEA

### PR DESCRIPTION
**Pull request by penguinland**
I recommend reviewing this commit-by-commit!

I want feedback now, before this goes further. The general change is:
- Create a new `DataReader` interface that provides a channel full of NMEA sentences from a device
- Create a `SerialDataReader` struct that implements this interface for a serial device
- Use the `SerialDataReader` in the `SerialNMEAMovementSensor`

Future changes I anticipate, once this is in:
- Create an `I2cDataReader` that also implements `DataReader` for an I2C device.
- Make `NewSerialGPSNMEA` and `NewPmtkI2CGPSNMEA` both return the same struct, except that they pass in different implementations of `DataReader` to that struct. Get rid of all the duplicated code between `gpsnmea/pmtI2C.go` and `gpsnmea/serial.go`

Tried on `piworld`: gps-nmea (serial) and gps-nmea-rtk-i2c appear to work normally. The gps-nmea-rtk-serial device is currently unplugged, but I'll double-check that still works later when it's plugged in again. I've also double-checked that it shuts down relatively quickly, too (no hanging on `Close()`)!

### CODE REVIEW

File: components/movementsensor/gpsnmea/serial.go [hunk 1]
```diff
@@ -95,33 +91,25 @@ func (g *SerialNMEAMovementSensor) Start(ctx context.Context) error {
 	g.activeBackgroundWorkers.Add(1)
 	utils.PanicCapturingGo(func() {
 		defer g.activeBackgroundWorkers.Done()
-		r := bufio.NewReader(g.dev)
 		for {
 			select {
 			case <-g.cancelCtx.Done():
 				return
-			default:
+			case line := <-g.dev.Lines():
```
Comment (penguinland) [inline components/movementsensor/gpsnmea/serial.go:98]: I wonder if the compiler is smart enough to just call `Lines()` once, or if it calls it each time through the loop. Maybe it should be moved into a variable before the loop starts, so there's no doubt.

File: components/movementsensor/gpsnmea/serial.go [hunk 2]
```diff
@@ -39,8 +37,7 @@ type SerialNMEAMovementSensor struct {
 	lastCompassHeading movementsensor.LastCompassHeading
 	isClosed           bool
 
-	dev  io.ReadWriteCloser
-	path string
+	dev DataReader
```
Comment (penguinland) [inline components/movementsensor/gpsnmea/serial.go:40]: A reminder as much for me as for other reviewers: we're still calling `dev.Close()` down in `SerialNMEAMovementSensor.Close()`. Although this PR doesn't touch that function, we _are_ correctly closing the data reader.

File: components/movementsensor/gpsnmea/serial.go [hunk 3]
```diff
@@ -95,33 +91,25 @@ func (g *SerialNMEAMovementSensor) Start(ctx context.Context) error {
 	g.activeBackgroundWorkers.Add(1)
 	utils.PanicCapturingGo(func() {
 		defer g.activeBackgroundWorkers.Done()
-		r := bufio.NewReader(g.dev)
 		for {
 			select {
```
Comment (penguinland) [inline components/movementsensor/gpsnmea/serial.go:95]: Having one select statement for everything isn't good enough: if multiple cases are satisfied, there's no guarantee on which one will fire, and perhaps we'll read lines forever without noticing that we should shut down. Have a second, non-blocking select statement for just the context.

**Comment by penguinland** [comment]
Tried on piworld again: the gps-nmea serial component takes a disconcertingly long time to get a satellite lock, but it does seem to work. and maybe that's normal? @susmitaSanyal will have a better sense of what's typical than I would...

File: components/movementsensor/gpsnmea/data_reader.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/jacobsa/go-serial/serial"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/logging"
+)
+
+// DataReader represents a way to get data from a GPS NMEA device. We can read data from it using
+// the channel in Lines, and we can close the device when we're done.
+type DataReader interface {
+	Lines() chan string
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsnmea/data_reader.go:18]: Let's change this from Lines to Messages

**Review by susmitaSanyal — APPROVED** [review]

**Pull request merged**